### ISSUE TITLE
We should support keywords in the datastore

### ISSUE THREAD

**Issue by Gustry** [inasafe/inasafe#3244]
### Problem
We should support keywords in the datastore for read and write with the monkey patching.

For now we have only two datastores : 
- folder : we should use the normal .xml file next to the shp or the tiff.
- geopackage : we should store the keywords in the geopackage itself.

We will add soon : 
- postgis

**Comment by timlinux** [comment]
Lets wait for QGIS 3 metadata in datasource support

**Comment by Gustry** [comment]
I just started a prototype in the PR above.

**Comment by mbernasocchi** [comment]
hi @timlinux I encoutered a bug related to this. When you load two different layers from the same geopackage they will share the keywords.
![screenshot from 2018-07-05 10-40-23](https://user-images.githubusercontent.com/233663/42295615-6e3bfc5e-8040-11e8-89ee-b6c74e6b9a45.png)
![screenshot from 2018-07-05 10-40-12](https://user-images.githubusercontent.com/233663/42295616-6e7846f0-8040-11e8-92a8-ef95ef5371b8.png)

Are you guys still working on @Gustry's PR #4921 ?

**Comment by Gustry** [comment]
You have this bug because InaSAFE doesn't care about what is after the `|` in your layer path unfortunately.
https://github.com/inasafe/inasafe/blame/master/safe/metadata/base_metadata.py#L196

So I guess you have a XML file sitting next to your geopackage. (one XML keywords file for one geopackage with many layers inside!)

Maybe it would be better to store keywords in the inasafe metadata DB?

BTW, is there not a kind of bug from QGIS? On the source text of your point layer, I would expect to have the `layername` property. How InaSAFE can know which layer it is in the geopackage? I already noticed this behavior.

**Comment by mbernasocchi** [comment]
> BTW, is there not a kind of bug from QGIS? On the source text of your point layer, I would expect to have the layername property. How InaSAFE can know which layer it is in the geopackage? I already noticed this behavior.
yes, I think this happens when you load a GPKG that has only one layer. after adding multiple layers probably it just picks the first layer.
@nyalldawson is it a bug or a feature? :)

**Comment by nyalldawson** [comment]
@mbernasocchi I'm a bit lost sorry, can you summarise exactly what the issue is?

**Comment by mbernasocchi** [comment]
when you two issues here the first is the one this bug report is about (keywords being per gpkg file and not per layer) but it is not the one I'm asking you about.

the second (The QGIS Question risen by @Gustry) is that when you add a GPKG with only one layer it is loaded as  `/path/file.gpkg`
when you add a further layer to this file it is loaded as:  `/path/file.gpkg|layername=SecondLayer` which is the behaviour I'd expect in every case also when adding a GPKG that has only one layer

**Comment by nyalldawson** [comment]
@mbernasocchi 

Seems like a bug to me too
